src/github.com/matrix-org/dendrite/clientapi/sync/requestpool.go

```diff
@@ -54,34 +75,70 @@ func (rp *RequestPool) OnIncomingSyncRequest(req *http.Request) util.JSONRespons
 		"timeout": timeout,
 	}).Info("Incoming /sync request")
 
-	res, err := rp.currentSyncForUser(syncReq)
-	if err != nil {
-		return httputil.LogThenError(req, err)
-	}
-	return util.JSONResponse{
-		Code: 200,
-		JSON: res,
+	// Fork off 2 goroutines: one to do the work, and one to serve as a timeout.
+	// Whichever returns first is the one we will serve back to the client.
+	// TODO: Currently this means that cpu work is timed, which may not be what we want long term.
+	timeoutChan := make(chan struct{})
+	timer := time.AfterFunc(timeout, func() {
+		close(timeoutChan) // signal that the timeout has expired
+	})
+
+	done := make(chan util.JSONResponse)
+	go func() {
+		syncData, err := rp.currentSyncForUser(syncReq)
+		timer.Stop()
+		var res util.JSONResponse
+		if err != nil {
+			res = httputil.LogThenError(req, err)
+		} else {
+			res = util.JSONResponse{
+				Code: 200,
+				JSON: syncData,
+			}
+		}
+		done <- res
+		close(done)
+	}()
+
+	select {
+	case <-timeoutChan: // timeout fired
+		return util.JSONResponse{
+			Code: 200,
+			JSON: []struct{}{}, // return empty array for now
+		}
+	case res := <-done: // received a response
+		return res
 	}
 }
 
-// OnNewEvent is called when a new event is received from the room server
-func (rp *RequestPool) OnNewEvent(ev *gomatrixserverlib.Event) {
+// OnNewEvent is called when a new event is received from the room server. Must only be
+// called from a single goroutine.
```

Comment:
might be worth noting why (presumably to avoid races between updates which end up in the wrong order?)